### PR TITLE
Check if failover target is ready for failover

### PR DESCRIPTION
**Pull request by ShyamsundarR**
Check if a VRG either exists as Primary on the failoverCluster
or if Secondary protectes PVCs via VolSync, in all other cases
Secondary cluster is not ready for a failover.

e.g a hub loss and a managed cluster loss when the surviving cluster
was the Secondary may not be a peer ready cluster to failover to.

Additionally:
Check if VRG has reached Primary state before cleanup

In cases where switchToCluster loops out due to errors or due to
waiting for VRG to manifest on the failoverCluster, the core failover
loop does not additionally check is VRG has reached the Primary state.

This commit addresses this gap.

NOTE: In such cases placement decision is also potentially not updated
but that is ensured before entering cleanup phase in ensureActionCompleted

### CODE REVIEW

**Comment by ShyamsundarR** [comment]
TODO:

- [ ] Complete e2e runs with VR and VS
- [ ] Add envtest for fake Secondary VRG to simulate failover failures

**Comment by ShyamsundarR** [comment]
@BenamarMk and @raghavendra-talur code is complete for review, will be doing e2e tests tomorrow.

File: controllers/drplacementcontrol.go [hunk 1]
```diff
@@ -367,6 +382,32 @@ func (d *DRPCInstance) RunFailover() (bool, error) {
 	return d.switchToFailoverCluster()
 }
 
+// isValidFailoverTarget determines if the passed in cluster is a valid target to failover to. A valid failover target
+// may already be Primary, if it is Secondary then it has to be protecting PVCs with VolSync.
+// NOTE: Currently there is a gap where, right after DR protection when a Secondary VRG is not yet created for VolSync
+// workloads, a failover if initiated will pass these checks. When we fix to retain VRG for VR as well, a more
+// deterministic check for VRG as Secondary can be performed.
+func (d *DRPCInstance) isValidFailoverTarget(cluster string) bool {
+	vrg, ok := d.vrgs[cluster]
```
Comment (BenamarMk) [inline controllers/drplacementcontrol.go:391]: This function may incorrectly declare the cluster as healthy due to failures in retrieving the VRG from it.

Comment (ShyamsundarR) [reply controllers/drplacementcontrol.go:391]: Good catch, only way out I see is to call `GetVRGFromManagedCluster` and ignore NotFound error and ensure we found a VRG or not. Stashed VRGs in DRPCInstance will not help with the error. Any other thoughts?

Comment (BenamarMk) [reply controllers/drplacementcontrol.go:391]: Maybe go the long way. Change the d.vrgs to be a
```
map[string]*VRGQuery
```
where the VRGQuery is type of
```
type VRGQuery struct {
  vrg *rmn.VolumeReplicationGroup
  requestResult bool
}
```
Where `requestResult` is either true or false. When true, it means successfully queried the cluster, the vrg pointer determines whether the VRG was found or not.  If false, it means the request failed and the `vrg` field is **guaranteed** to be `nil`.

I am not sure if you want to do that. However, in the long term, we should actually refrain from calling `GetVRGFromManagedCluster` from all over the place (unfortunately we do today).  Each reconciliation we populate `d.args` and we should use that array throughout.

Comment (ShyamsundarR) [reply controllers/drplacementcontrol.go:391]: Agreed, for now I am going to use the shortcut (the other change is quite intrusive) and follow up with a broader change to leverage d.vrgs with the required information.

**Review by BenamarMk — APPROVED** [review]

**Review by BenamarMk — APPROVED** [review]
LGTM

**Comment by ShyamsundarR** [comment]
@BenamarMk e2e against Volsync and volrep completed (series of deploy->failover->relocate->failover like scenarios), no issues found with the changes.

File: controllers/drplacementcontrol.go (hunk 1 above)
Comment (ShyamsundarR) [reply controllers/drplacementcontrol.go]: Issue to track using cached VRGs opened here: https://github.com/RamenDR/ramen/issues/1384

**Pull request merged**